src/main/java/com/microsoft/azure/relay/ClientWebSocket.java

```diff
@@ -173,36 +174,41 @@ boolean isOpen() {
 	 * @throws TimeoutException thrown when a complete message frame is not received within the timeout.
 	 */
 	CompletableFuture<ByteBuffer> readBinaryAsync(Duration timeout) {
-		
-		return CompletableFutureUtil.timedSupplyAsync(timeout, () -> {
-				LinkedList<byte[]> fragments = new LinkedList<byte[]>();
-				boolean receivedWholeMsg;
-				int messageSize = 0;
-				
-				do {
-					MessageFragment fragment = fragmentQueue.dequeueAsync().join();
-					if (fragment == null) {
-						break;
-					}				
-					messageSize += fragment.getBytes().length;
-					fragments.add(fragment.getBytes());
-					receivedWholeMsg = fragment.isEnd();
-				} 
-				while (!receivedWholeMsg);
-	
-				byte[] message = new byte[messageSize];
-				int offset = 0;
-				for (byte[] bytes : fragments) {
-					System.arraycopy(bytes, 0, message, offset, bytes.length);
-					offset += bytes.length;
+		LinkedList<byte[]> fragments = new LinkedList<byte[]>();
+		AtomicInteger messageSize = new AtomicInteger(0);
+		return readFragmentsAsync(fragments, messageSize, timeout).thenApply((nullResult) -> {
+			byte[] message = new byte[messageSize.get()];
+			int offset = 0;
+			for (byte[] bytes : fragments) {
+				System.arraycopy(bytes, 0, message, offset, bytes.length);
+				offset += bytes.length;
+			}
+			RelayLogger.logEvent("receivedBytes", this, String.valueOf(message.length));
+			return ByteBuffer.wrap(message);
+		});
+	}
+
+	private CompletableFuture<Void> readFragmentsAsync(LinkedList<byte[]> fragments, AtomicInteger messageSize, Duration timeout) {
+		TimeoutHelper timeoutHelper = (timeout != null) ? new TimeoutHelper(timeout, true) : null;
+		return fragmentQueue.dequeueAsync(timeout).thenCompose(fragment -> {
+			if (fragment == null) {
+				return CompletableFuture.completedFuture(null);
+			}
+
+			messageSize.getAndAdd(fragment.getBytes().length);
+			fragments.add(fragment.getBytes());
+
+			if (!fragment.isEnd()) {
+				if (timeout != null) {
+					return readFragmentsAsync(fragments, messageSize, timeoutHelper.remainingTime());
+				} else {
+					return readFragmentsAsync(fragments, messageSize, null);
 				}
```

Comment:
It's largely a personal preference but since this method is already using a trick (which is cool) I would try to keep the number of possible re-entry points as little as possible. Look at this alternate then use whichever you prefer.

    Duration remainingTime = null; // Or use ? : 
    if (timeout != null) {
        remainingTime = timeoutHelper.remainingTime();
    }
    return readFragmentsAsync(fragments, messageSize, remainingTime);

Reply:
This makes me wonder, in Java-land would it be better if TimeoutHelper allowed a null Duration?

---
In reply to: [257368397](https://github.com/Azure/azure-relay-java/pull/14#discussion_r257368397) [](ancestors = 257368397)

Reply:
I modified TimeoutHelper so that if a null duration was passed in it would be interpreted as max duration

---
In reply to: [257370849](https://github.com/Azure/azure-relay-java/pull/14#discussion_r257370849) [](ancestors = 257370849,257368397)